src/main/java/jenkins/scm/api/SCMSource.java

```diff
@@ -609,13 +609,33 @@ protected SCMRevision retrieve(@NonNull SCMHead head, @NonNull TaskListener list
      * @throws IOException if an error occurs while performing the operation.
      * @throws InterruptedException if any thread has interrupted the current thread.
      * @since 1.3
+     * @deprecated rather call {@link #fetch(String, TaskListener, Item)}
      */
+    @Deprecated
     @CheckForNull
     public final SCMRevision fetch(@NonNull String thingName, @CheckForNull TaskListener listener)
             throws IOException, InterruptedException {
         return retrieve(thingName, defaultListener(listener));
     }
 
+    /**
+     * Looks up a specific thingName based on some SCM-specific set of permissible syntaxes.
+     * Delegates to {@link #retrieve(String, TaskListener)}.
+     * @param thingName might be a branch name, a tag name, a cryptographic hash, a change request number, etc.
+     * @param listener the task listener (optional)
+     * @param context an associated context to supersede {@link #getOwner}, such as a job in which this is running
+     * @return a valid {@link SCMRevision} corresponding to the argument, with a usable corresponding head, or
+     * {@code null} if malformed or not found
+     * @throws IOException if an error occurs while performing the operation.
+     * @throws InterruptedException if any thread has interrupted the current thread.
+     * @since 2.5
+     */
+    @CheckForNull
+    public final SCMRevision fetch(@NonNull String thingName, @CheckForNull TaskListener listener, @CheckForNull Item context)
```

Comment:
If this change makes sense for this API entry point then it makes sense for all.

That is:

```
    protected abstract void retrieve(@CheckForNull SCMSourceCriteria criteria,
                            @NonNull SCMHeadObserver observer,
                            @CheckForNull SCMHeadEvent<?> event,
                            @NonNull TaskListener listener)
            throws IOException, InterruptedException;
```

should probably be moved to a variant that includes the context item. That is the main implementation point. 

The idea is this: the API provides some well-known use cases, such as getting a specific revision of some specific thing. Implementations do not need to implement for all these use cases, instead they can start by implementing the main SPI points (i.e. the `abstract` ones) and they will get an implementation of the common use cases "for free"... of course the free implementation is not optimized, so implementations can improve things by overriding the various SPI optimization points.

What you are doing with this change is addressing your use case only, not the generic need. As such this change in its current format would:

1. Require all implementations to provide a context aware SPI override of a hidden method (it's non-abstract)
2. Not actually work against implementations that do not provide the hidden method.

Reply:
@stephenc 
What you're describing sound like a significantly different design and of larger scope as well.  

Perhaps that would be separate JIRA issue to be done later?